File: euler.c
// date 20/04/ 2020
// euler's method for solving differential equation
// diff eqn : y' = t-t*t +1  ; 0<= t <=2    ,y(0) =0.5

#include<stdio.h>
#include<math.h>

float func(float t,float y)
{
  float f;
  f= t-t*t +1;                // r.h.s. of differential equation
  return f;
}

float funct(float t,float y)
{ float g;
  g= (t*t)/2-(t*t*t)/3 +t + 0.5;
  return g;
}

void main()
{
  float t,y,h,t_upper,k,l,diff;
  
  h=0.2;
  t=0;                                // initial value of t
  y=0.5 ;                             // since y(0)=0.5
  t_upper=2;                           // upper limit of t

  while(t<= t_upper)
    {
      l=funct(t,y);
      diff=l-y;                          // difference b/w exact and numerical sol
      k= h*func(t,y);
      y=y+k;
      t=t+h;
      printf("%0.3f\t %0.3f\t %0.3f\n",t,y,diff);
      
    }
}

File: euler.py
#  date 11/04/2020
#  Diff. equation : dy/dt = y/t - (y/t)**2  ,  1 <= t <= 2

import numpy as np
import matplotlib.pyplot as plt

h=0.1                               # distance between mess points
t=np.arange(1,2.1,h)                # an array of mess points
t_0=1
y_1=1                               # initial condition
y=[y_1]                             # solution array

def func(u,v):
    return(h*(u/v)*(1-(u/v)))       # h*dy/dt

for i in range(len(t)-1):          
    k= y[i] + func(y[i],t[i])       # euler method's approximated value at alternate mess points
    y.append(k)                     # updating solution arange

y_exact= t/(1+np.log(t))            # exact solution

plt.plot(t,y,'b',label='numerically approximated solution')
plt.plot(t,y_exact,'r',label='exact solution')
plt.title("solving by euler's method ")
plt.xlabel('t')
plt.ylabel('y')
plt.legend()
plt.show()

error =0                             # initiating the error term
for i in range(len(t)):
    error=error+(y_exact[i]-y[i])    # error in corresponding terms

absolut_error=(np.absolute(error))/11     # 11 mess points

erro_r=0
for i in range(len(t)):
    erro_r=erro_r + ((y_exact[i]-y[i])/y_exact[i])

relat_error=(np.absolute(erro_r))/11

print('absolute error is :', absolut_error)
print('relative error is :', relat_error)
